### PR TITLE
 Add TORCH_CHECK to disable sub for bool tensors

### PR DESCRIPTION
**Pull request by xuhdev**
This resolves two issues in one shot:

- sub shouldn't be available for bool type.
- When sub is applied to an unsupported type, the current error messages
  shows "add_cpu/add_cuda is not implemented for [type]". They should be
  "sub_cpu/sub_cuda" instead.

### CODE REVIEW

**Comment by xuhdev** [comment]
Same as #23335 , but corrected error message for bool-bool

File: test/test_torch.py [hunk 1]
```diff
@@ -1706,6 +1706,14 @@ def test_add(self):
             expected = torch.zeros(2, 3, device=device).bool()
             self.assertEqual(res, expected)
 
+    def test_bool_sub(self):
+        for device in torch.testing.get_all_device_types():
+            m1 = torch.tensor([True, False, False, True, False, False], dtype=torch.bool, device=device)
+            m2 = torch.tensor([True, True, False, False, False, True], dtype=torch.bool, device=device)
+            self.assertRaises(RuntimeError, lambda: m1 - m2)
```
Comment (gchanan) [inline test/test_torch.py]: nit: can we check we get the correct error message in each case?

**Review by gchanan — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/BinaryOps.cpp [hunk 2]
```diff
@@ -100,7 +100,17 @@ Tensor& mul_(Tensor& self, const Tensor& other) {
   return native::mul_out(self, self, other);
 }
 
+// Basic checking for all sub functions.
+static inline void sub_check(const Tensor& self, const Tensor& other) {
+  TORCH_CHECK(self.scalar_type() != kBool || other.scalar_type() != kBool,
```
Comment (gchanan) [inline aten/src/ATen/native/BinaryOps.cpp]: I suggested slightly different wording on the old PR, can you move that over here?

Comment (xuhdev) [reply aten/src/ATen/native/BinaryOps.cpp]: Oops, just saw that. Let me change here.

**Comment by xuhdev** [comment]
@gchanan  This PR should be ready now

**Review by gchanan — APPROVED** [review]

**Comment by gchanan** [comment]
thanks @xuhdev!

**Comment by gchanan** [comment]
@izdeby can you handle landing this?  I don't know if there are complications with the other Boolean--related PRs you are working on.

**Comment by xuhdev** [comment]
Thank you all for your review!

**Pull request closed without merging**